### PR TITLE
Moved some implementations into the abstract base class

### PR DESCRIPTION
**Pull request by PhillipVH**
Did some house cleaning.

### CODE REVIEW

File: src/main/java/za/ac/sun/cs/search/singleagent/Board.java [hunk 1]
```diff
@@ -137,9 +144,45 @@ public String toString() {
...
+                return makeMove(Direction.RIGHT);
+            case RIGHT:
+                return makeMove(Direction.LEFT);
+            default:
+                return null;
+        }
     }
 
-    public abstract Board makeMove(Direction move);
+    public Board makeMove(Direction move) {
```
Comment (PhillipVH) [inline src/main/java/za/ac/sun/cs/search/singleagent/Board.java]: `makeMove` and `swapTiles` are now in the abstract class, just means we need to explicitly cast them to either an `ImplictBoard` or `ExplicitBoard` where appropriate.

File: src/main/java/za/ac/sun/cs/search/singleagent/IDAStarAgent.java [hunk 2]
```diff
@@ -35,7 +33,7 @@ public IDAStarAgent(short[] configuration) {
             bound = i;
         }               
 
-        return path.toArray(new Direction[path.size()]);
+        return path.toArray(new Direction[0]);
```
Comment (PhillipVH) [inline src/main/java/za/ac/sun/cs/search/singleagent/IDAStarAgent.java]: `toArray` [apparently](https://docs.oracle.com/javase/7/docs/api/java/util/Vector.html#toArray) figures out size for us.

File: src/main/java/za/ac/sun/cs/search/singleagent/IDAStarAgent.java [hunk 3]
```diff
@@ -25,7 +23,7 @@ public IDAStarAgent(short[] configuration) {
      */
     @Override
     public Direction[] solve() {
-        Stack<Direction> path = new Stack<Direction>();
+        Stack<Direction> path = new Stack<>();
```
Comment (PhillipVH) [inline src/main/java/za/ac/sun/cs/search/singleagent/IDAStarAgent.java]: You might want to consider `ArrayDeque`, as stated on the [`Stack` documentation](https://docs.oracle.com/javase/7/docs/api/java/util/Stack.html): 
> A more complete and consistent set of LIFO stack operations is provided by the Deque interface and its implementations, which should be used in preference to this class.

Comment (Zhunaidm) [reply src/main/java/za/ac/sun/cs/search/singleagent/IDAStarAgent.java]: I'll look into it. Both provide the functionality needed.

File: src/main/java/za/ac/sun/cs/search/singleagent/Board.java (hunk 1 above)
Comment (Zhunaidm) [reply src/main/java/za/ac/sun/cs/search/singleagent/Board.java]: Sure.

File: src/main/java/za/ac/sun/cs/search/singleagent/IDAStarAgent.java (hunk 2 above)
Comment (Zhunaidm) [reply src/main/java/za/ac/sun/cs/search/singleagent/IDAStarAgent.java]: Cool, did not know that.

**Pull request merged**